# Model building for house price index

In [39]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [40]:
df = pd.read_csv('CSUSHPISA.csv')
df.head()

DATE  Year  CSUSHPISA  CSUSHPISA Average  GDP growth  \
0  1/1/2003  2003    128.461         133.731333        0.03   
1  2/1/2003  2003    129.355                NaN         NaN   
2  3/1/2003  2003    130.148                NaN         NaN   
3  4/1/2003  2003    130.884                NaN         NaN   
4  5/1/2003  2003    131.735                NaN         NaN   

   Unemployment rates  inflation  Interest Rates  Population growth  \
0                0.06       0.02            0.04               0.01   
1                 NaN        NaN             NaN                NaN   
2                 NaN        NaN             NaN                NaN   
3                 NaN        NaN             NaN                NaN   
4                 NaN        NaN             NaN                NaN   

   net migration rate  
0                 3.7  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN

In [41]:
#handling missing values and delete columns which will not needed 
df = df.drop(['DATE', 'CSUSHPISA'], axis=1)
df = df.dropna()
df

Year  CSUSHPISA Average  GDP growth  Unemployment rates  inflation  \
0    2003         133.731333        0.03                0.06       0.02   
12   2004         150.440250        0.04                0.06       0.03   
24   2005         171.737000        0.03                0.05       0.03   
36   2006         183.447500        0.03                0.05       0.03   
48   2007         179.918917        0.02                0.05       0.03   
60   2008         164.057417        0.00                0.06       0.04   
72   2009         148.545083       -0.03                0.09       0.00   
84   2010         144.674500        0.03                0.10       0.02   
96   2011         139.259500        0.02                0.09       0.03   
108  2012         140.993917        0.02                0.08       0.02   
120  2013         154.520750        0.02                0.07       0.01   
132  2014         164.698417        0.02                0.06       0.02   
144  2015         172.181833        0.03                0.05       0.00   
156  2016         180.925583        0.02                0.05       0.01   
168  2017         191.398000        0.02                0.04       0.02   
180  2018         202.477750        0.03                0.04       0.02   
192  2019         209.465167        0.02                0.04       0.02   
204  2020         222.145833       -0.03                0.08       0.01   
216  2021         260.049250        0.06                0.05       0.05   
228  2022         298.488333        0.02                0.04       0.08   
240  2023         302.066875        0.05                0.04       0.04   

     Interest Rates  Population growth  net migration rate  
0              0.04               0.01               3.700  
12             0.04               0.01               3.679  
24             0.04               0.01               3.658  
36             0.05               0.01               3.637  
48             0.05               0.01               3.616  
60             0.04               0.01               3.595  
72             0.03               0.01               3.506  
84             0.03               0.01               3.417  
96             0.03               0.01               3.329  
108            0.02               0.01               3.240  
120            0.02               0.01               3.151  
132            0.03               0.01               3.107  
144            0.02               0.01               3.062  
156            0.02               0.01               3.018  
168            0.02               0.01               2.973  
180            0.03               0.01               2.929  
192            0.02               0.01               2.893  
204            0.01               0.00               2.857  
216            0.02               0.00               2.820  
228            0.03               0.00               2.784  
240            0.04               0.01               2.748

In [42]:
df.describe()

Year  CSUSHPISA Average  GDP growth  Unemployment rates  \
count    21.000000          21.000000   21.000000           21.000000   
mean   2013.000000         186.439200    0.021429            0.059524   
std       6.204837          48.718257    0.021044            0.018568   
min    2003.000000         133.731333   -0.030000            0.040000   
25%    2008.000000         150.440250    0.020000            0.050000   
50%    2013.000000         172.181833    0.020000            0.050000   
75%    2018.000000         202.477750    0.030000            0.070000   
max    2023.000000         302.066875    0.060000            0.100000   

       inflation  Interest Rates  Population growth  net migration rate  
count  21.000000       21.000000          21.000000           21.000000  
mean    0.025238        0.030000           0.008571            3.224714  
std     0.017782        0.010954           0.003586            0.337434  
min     0.000000        0.010000           0.000000            2.748000  
25%     0.020000        0.020000           0.010000            2.929000  
50%     0.020000        0.030000           0.010000            3.151000  
75%     0.030000        0.040000           0.010000            3.595000  
max     0.080000        0.050000           0.010000            3.700000

In [43]:
df.columns

Index(['Year', 'CSUSHPISA Average', 'GDP growth', 'Unemployment rates',
       'inflation', 'Interest Rates', 'Population growth',
       'net migration rate'],
      dtype='object')

In [44]:
# Select the independent variables
X = df[['GDP growth','Unemployment rates', 'Interest Rates', 'inflation','Population growth', 'net migration rate']]

# Define the dependent variable
y = df['CSUSHPISA Average']

# Multiple Regression 

In [45]:
# Split the data into a training set and a testing set
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Create and train the multiple regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)

Mean Squared Error: 740.4331121617663
R-squared: 0.7871006072075525


# Polynomial Regression

In [48]:
from sklearn.preprocessing import PolynomialFeatures

# Create polynomial features
poly = PolynomialFeatures(degree=2)  # can be adjusted the degree as needed
X_poly = poly.fit_transform(X)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_poly, y, test_size=0.2, random_state=0)

# Create and train the polynomial regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)

Mean Squared Error: 146.31769053615875
R-squared: 0.9579287487846225


# Decision Tree

In [47]:
from sklearn.tree import DecisionTreeRegressor
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=0)

# Create and train the decision tree regression model
model = DecisionTreeRegressor()
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)

print("Mean Squared Error:", mse)
print("R-squared:", r2)

Mean Squared Error: 186.67170871758327
R-squared: 0.9463256129625665


Achieving high R-squared values, particularly in the range of 0.94 to 0.96, is a strong indication the models are effectively explaining the variations in home prices.

The substantial increase in R-squared values for both the decision tree and polynomial regression models demonstrates that they are now providing a highly accurate representation of the relationships between the included factors and home prices.

Both the decision tree and polynomial regression models have performed impressively.